cpp/src/tsne/fft_tsne.cuh

```diff
@@ -0,0 +1,527 @@
+#include <raft/linalg/eltwise.cuh>
+#include <raft/mr/device/buffer.hpp>
+#include <raft/stats/sum.cuh>
+#include <rmm/device_vector.hpp>
+#include "fft_kernels.cuh"
+#include "utils.cuh"
+
+namespace ML {
+namespace TSNE {
+
+const static int NTHREADS_1024 = 1024;
+const static int NTHREADS_128 = 128;
+const static int NTHREADS_32 = 32;
+
+struct FunctionalSqrt {
+  template <typename value_t>
+  __host__ __device__ float operator()(const value_t &x) const {
+    return pow(x, 0.5);
+  }
+};
+struct FunctionalSquare {
+  template <typename value_t>
+  __host__ __device__ float operator()(const value_t &x) const {
+    return x * x;
+  }
+};
+
+template <typename T>
+cufftResult CUFFTAPI cufft_MakePlanMany(cufftHandle plan, T rank, T *n,
+                                        T *inembed, T istride, T idist,
+                                        T *onembed, T ostride, T odist,
+                                        cufftType type, T batch,
+                                        size_t *workSize);
+
+cufftResult CUFFTAPI cufft_MakePlanMany(cufftHandle plan, int rank, int64_t *n,
+                                        int64_t *inembed, int64_t istride,
+                                        int64_t idist, int64_t *onembed,
+                                        int64_t ostride, int64_t odist,
+                                        cufftType type, int64_t batch,
+                                        size_t *workSize) {
+  return cufftMakePlanMany64(
+    plan, rank, reinterpret_cast<long long int *>(n),
+    reinterpret_cast<long long int *>(inembed),
+    static_cast<long long int>(istride), static_cast<long long int>(idist),
+    reinterpret_cast<long long int *>(onembed),
+    static_cast<long long int>(ostride), static_cast<long long int>(odist),
+    type, static_cast<long long int>(batch), workSize);
+}
+cufftResult CUFFTAPI cufft_MakePlanMany(cufftHandle plan, int rank, int *n,
+                                        int *inembed, int istride, int idist,
+                                        int *onembed, int ostride, int odist,
+                                        cufftType type, int batch,
+                                        size_t *workSize) {
+  return cufftMakePlanMany(plan, rank, n, inembed, istride, idist, onembed,
+                           ostride, odist, type, batch, workSize);
+}
+
+template <typename value_t, typename value_idx>
+std::pair<value_t, value_t> min_max(const value_t *Y, const value_idx n,
+                                    cudaStream_t stream, int iter) {
+  value_t min_h, max_h;
+
+  rmm::device_uvector<value_t> min_d(1, stream);
+  rmm::device_uvector<value_t> max_d(1, stream);
+
+  min_d.set_element(0, std::numeric_limits<value_t>::max(), stream);
+  max_d.set_element(0, std::numeric_limits<value_t>::lowest(), stream);
+
+  raft::update_host(&min_h, min_d.data(), 1, stream);
+  raft::update_host(&max_h, max_d.data(), 1, stream);
+
+  auto nthreads = 1024;
```

Comment:
Probably want to use 256 for this one, right?

Reply:
Since it's a block reduce, don't we want the blocks to be large to minimize atomic writes?